pageserver/benches/bench_layer_map.rs

```diff
@@ -243,63 +245,117 @@ fn bench_from_captest_env(c: &mut Criterion) {
 fn bench_sequential(c: &mut Criterion) {
-    let mut layer_map = LayerMap::default();
-
     // Init layer map. Create 100_000 layers arranged in 1000 diagonal lines.
     //
     // TODO This code is pretty slow and runs even if we're only running other
     //      benchmarks. It needs to be somewhere else, but it's not clear where.
     //      Putting it inside the `bench_function` closure is not a solution
     //      because then it runs multiple times during warmup.
     let now = Instant::now();
+    let mut layer_map = LayerMap::default();
     for i in 0..100_000 {
-        // TODO try inserting a super-wide layer in between every 10 to reflect
-        //      what often happens with L1 layers that include non-rel changes.
-        //      Maybe do that as a separate test.
         let i32 = (i as u32) % 100;
         let zero = Key::from_hex("000000000000000000000000000000000000").unwrap();
         let layer = DummyImage {
             key_range: zero.add(10 * i32)..zero.add(10 * i32 + 1),
-            lsn: Lsn(10 * i),
+            lsn: Lsn(i),
         };
         layer_map.insert_historic(Arc::new(layer));
     }
+    println!("Finished layer map init in {:?}", now.elapsed());
 
-    // Manually measure runtime without criterion because criterion
-    // has a minimum sample size of 10 and I don't want to run it 10 times.
-    println!("Finished init in {:?}", now.elapsed());
+    // Init bst layer map with the same layers
+    let now = Instant::now();
+    let mut bstlm = PersistentLayerMap::new();
+    let mut sorted_layers: Vec<_> = layer_map.iter_historic_layers().collect();
+    sorted_layers.sort_by(|a, b| a.get_lsn_range().start.cmp(&b.get_lsn_range().start));
+    // TODO implement out of order inserts
+    for layer in sorted_layers {
+        let kr = layer.get_key_range();
+        let lr = layer.get_lsn_range();
+
+        bstlm.insert(
+            kr.start.to_i128(),
+            kr.end.to_i128(),
+            lr.start.0,
+            format!("Layer {}", lr.start.0),
+        );
+    }
+    println!("Finished bst init in {:?}", now.elapsed());
 
     // Choose 100 uniformly random queries
     let rng = &mut StdRng::seed_from_u64(1);
     let queries: Vec<(Key, Lsn)> = uniform_query_pattern(&layer_map)
-        .choose_multiple(rng, 1)
+        .choose_multiple(rng, 100)
         .copied()
         .collect();
 
     // Define and name the benchmark function
-    c.bench_function("sequential_uniform_queries", |b| {
-        // Run the search queries
+    let mut group = c.benchmark_group("sequential_uniform_queries");
+    group.bench_function("current_code", |b| {
         b.iter(|| {
             for q in queries.clone().into_iter() {
                 layer_map.search(q.0, q.1).unwrap();
             }
         });
     });
+    group.bench_function("persistent_bst", |b| {
+        b.iter(|| {
+            for q in queries.clone().into_iter() {
+                bstlm.query(q.0.to_i128(), q.1 .0);
```

Comment:
Might be a good idea to have an unwrap here as well (compared to "current_code" above), also `criterion::black_box` for both/all. Unsure what happens to this as is, failed to find it from asm even after adding a `#[inline(never)]`, but that's probably me just not seeing it.

Tried to run these to compare, but it seems all of these fail with panic at `thread 'main' panicked at 'unexpected retroactive insert', pageserver/src/tenant/bst_layer_map.rs:55:17`, as is probably expected.

Reply:
Good idea! Created a follow-up issue